### PR TITLE
[luci-interpreter] Support Quantize op in luci-interpreter

### PR DESCRIPTION
**Pull request by BalyshevArtem**
This pr adds support for Quantize op in luci-interpreter.

issue: #8372

ONE-DCO-1.0-Signed-off-by: Artem Balyshev a.balyshev@partner.samsung.com

### CODE REVIEW

File: compiler/luci-interpreter/src/kernels/Quantize.test.cpp [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+  kernel.configure();
+  _memory_manager->allocate_memory(output_tensor);
+  kernel.execute();
+
+  EXPECT_THAT(extractTensorData<int8_t>(output_tensor),
+              ::testing::ElementsAreArray(ref_output_data));
+  EXPECT_THAT(extractTensorShape(output_tensor), ::testing::ElementsAreArray({1, 1, 2, 5}));
+}
+
+TEST_F(QuantizeTest, InvalidInputTypeNeg)
```
Comment (jinevening) [inline compiler/luci-interpreter/src/kernels/Quantize.test.cpp:179]: Negative tests should have `_NEG` postfix. This is for our SE tool to correctly collect statistics.

```suggestion
TEST_F(QuantizeTest, InvalidInputType_NEG)
```

Comment (jinevening) [reply compiler/luci-interpreter/src/kernels/Quantize.test.cpp:179]: Please update the below tests as well.

Comment (BalyshevArtem) [reply compiler/luci-interpreter/src/kernels/Quantize.test.cpp]: Thank you! Done :)

File: compiler/luci-interpreter/src/kernels/Quantize.cpp [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+      compute_shift_multiplier(static_cast<double>(input()->scale()),
+                               static_cast<double>(output()->scale()), &multiplier, &shift);
+
+      const auto input_shape = getTensorShape(input());
+      const auto output_shape = getTensorShape(output());
+
+      const auto input_data = getTensorData<int16_t>(input());
+
+      const auto size = tflite::MatchingFlatSize(input_shape, output_shape);
```
Comment (m-bronnikov) [inline compiler/luci-interpreter/src/kernels/Quantize.cpp:122]: I think we could extract this lines (except getTensorData<int16>) from swith-case in order to reduce code duplication.

File: compiler/luci-interpreter/src/kernels/Quantize.cpp [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+                                           input()->zero_point(), output()->zero_point(),
+                                           getTensorData<int16_t>(output()));
+          break;
+        }
+        case loco::DataType::S32:
+        {
+          luci_interpreter_pal::Requantize(input_data, size, multiplier, shift,
+                                           input()->zero_point(), output()->zero_point(),
+                                           getTensorData<int32_t>(output()));
+          break;
```
Comment (m-bronnikov) [inline compiler/luci-interpreter/src/kernels/Quantize.cpp:145]: We are support `S32` quantization? =0
I think better to remove this.

**Review by m-bronnikov — COMMENTED** [review]
Code itself looks good to me, but I think it would be nice to reduce code duplication and support S8 (U8) -> S16 quantization. 

For each input datatype we introduce switch-case block: huge amount of code which could be reduced.

For this purpose I suggest following helper:

```cpp
template <typename input_dtype>
void call_requantize_node(const Tensor *input, const Tensor *output)
{
  int32_t multiplier;
  int32_t shift;
  compute_shift_multiplier(input->scale(), output->scale(), &multiplier, &shift);

  const auto input_shape = getTensorShape(input);
  const auto output_shape = getTensorShape(output);
  const auto size = tflite::MatchingFlatSize(input_shape, output_shape);

  const auto input_data = getTensorData<input_dtype>(input);

  switch (output->element_type())
  {
    case loco::DataType::S8:
      luci_interpreter_pal::Requantize(input_data, size, multiplier, shift,
                                       input->zero_point(), output->zero_point(),
                                       getTensorData<int8_t>(output));
      break;
    case loco::DataType::U8:
      luci_interpreter_pal::Requantize(input_data, size, multiplier, shift,
                                       input->zero_point(), output->zero_point(),
                                       getTensorData<uint8_t>(output));
      break;
    case loco::DataType::S16:
      luci_interpreter_pal::Requantize(input_data, size, multiplier, shift,
                                       input->zero_point(), output->zero_point(),
                                       getTensorData<int16_t>(output));
      break;
    default:
      throw std::runtime_error("Unsupported quantized type, yet!");
  }
}
```

So, for each quantized input datatype we just can call this helper instead own switch-case block introducing.

File: compiler/luci-interpreter/src/kernels/Quantize.cpp (hunk 3 above)
Comment (BalyshevArtem) [reply compiler/luci-interpreter/src/kernels/Quantize.cpp]: Done

File: compiler/luci-interpreter/src/kernels/Quantize.cpp [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+      LUCI_INTERPRETER_CHECK(input()->zero_point() == 0 && output()->zero_point() == 0);
+    }
+  }
+  else if (input()->element_type() == loco::DataType::S8 ||
+           input()->element_type() == loco::DataType::U8)
+  {
+    // TODO: support S8 (U8) -> S16
+    LUCI_INTERPRETER_CHECK(output()->element_type() == loco::DataType::S8 ||
+                           output()->element_type() == loco::DataType::U8);
+  }
```
Comment (m-bronnikov) [inline compiler/luci-interpreter/src/kernels/Quantize.cpp:89]: Why `// TODO: support S8 (U8) -> S16`? As far as I understand we are support this requantization for now with `call_requantize` helper. IIUC we need just to extend `LUCI_INTERPRETER_CHECK`'` condition.

Comment (m-bronnikov) [reply compiler/luci-interpreter/src/kernels/Quantize.cpp:89]: Also, how about switch-case here for `input->element_type()`?

Comment (BalyshevArtem) [reply compiler/luci-interpreter/src/kernels/Quantize.cpp:89]: > Why `// TODO: support S8 (U8) -> S16`? As far as I understand we are support this requantization for now with `call_requantize` helper. IIUC we need just to extend `LUCI_INTERPRETER_CHECK`'` condition.

I am not sure Quantize kernel can work properly with such quantization due to tflite Quantize kernel doesn't support such quantization:   https://github.com/tensorflow/tensorflow/blob/acddd82daca0c2c3e374fb8ed9607a5771918c2e/tensorflow/lite/kernels/quantize.cc#L126
```c++
   // Requantize use case.
    if (input->type == kTfLiteInt16) {
      TF_LITE_ENSURE(context, output->type == kTfLiteInt8 ||
                                  output->type == kTfLiteInt16 ||
                                  output->type == kTfLiteInt32);
    } else if (input->type == kTfLiteInt32) {
      TF_LITE_ENSURE(
          context, output->type == kTfLiteInt8 || output->type == kTfLiteInt16);
    } else {
      TF_LITE_ENSURE(context,
                     input->type == kTfLiteInt8 || input->type == kTfLiteUInt8);
      TF_LITE_ENSURE(
          context, output->type == kTfLiteUInt8 || output->type == kTfLiteInt8);
    }
```

Comment (BalyshevArtem) [reply compiler/luci-interpreter/src/kernels/Quantize.cpp:89]: But If you think It will be ok, I will add this quantization case.

Comment (m-bronnikov) [reply compiler/luci-interpreter/src/kernels/Quantize.cpp:89]: I checked `tflite::Requantize` impl and looks like it works correctly for any quantized datatype depending only on `dtype`'s min/max and quantization zero point:
https://github.com/tensorflow/tensorflow/blob/acddd82daca0c2c3e374fb8ed9607a5771918c2e/tensorflow/lite/kernels/internal/reference/requantize.h#L51-L62

Comment (BalyshevArtem) [reply compiler/luci-interpreter/src/kernels/Quantize.cpp]: Done

File: compiler/luci-interpreter/src/kernels/Quantize.cpp [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+      break;
+    }
+    case loco::DataType::S8:
+    case loco::DataType::U8:
+    {
+      LUCI_INTERPRETER_CHECK(output()->element_type() == loco::DataType::S8 ||
+                             output()->element_type() == loco::DataType::U8 ||
+                             output()->element_type() == loco::DataType::S16);
+      break;
+    }
```
Comment (m-bronnikov) [inline compiler/luci-interpreter/src/kernels/Quantize.cpp:95]: 1. input dtype `S16` means `input->zero_point()` is `0` for any output dtype.
2. output dtype `S16` means 'output->zero_point()` is `0` for any input dtype.

```suggestion
    case loco::DataType::S16:
        LUCI_INTERPRETER_CHECK(input()->zero_point() == 0 && output()->zero_point() == 0);
    case loco::DataType::S8:
    case loco::DataType::U8:
    {
      LUCI_INTERPRETER_CHECK(output()->element_type() == loco::DataType::S8 ||
                             output()->element_type() == loco::DataType::U8 ||
                             output()->element_type() == loco::DataType::S16);
      if (output()->element_type() == loco::DataType::S16)
      {
        LUCI_INTERPRETER_CHECK(output()->zero_point() == 0);
      }
      break;
    }
```

**Review by binarman — APPROVED** [review]
LGTM, see comments

File: compiler/luci-interpreter/src/kernels/Quantize.cpp [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+  {
+    case loco::DataType::FLOAT32:
+    {
+      LUCI_INTERPRETER_CHECK(output()->element_type() == loco::DataType::U8 ||
+                             output()->element_type() == loco::DataType::S8 ||
+                             output()->element_type() == loco::DataType::S16);
+      break;
+    }
+    case loco::DataType::S16:
+      LUCI_INTERPRETER_CHECK(input()->zero_point() == 0);
```
[the remark this answers is not in the document]
Comment (binarman) [reply compiler/luci-interpreter/src/kernels/Quantize.cpp:79]: Yes, if you remove this line, it should compile fine.
Move this check before switch, this should help.

File: compiler/luci-interpreter/pal/cmsisnn/PALQuantize.h [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef LUCI_INTERPRETER_PAL_QUANTIZE_H
+#define LUCI_INTERPRETER_PAL_QUANTIZE_H
+
+#include "tensorflow/lite/kernels/internal/reference/reference_ops.h"
+#include "tensorflow/lite/kernels/internal/reference/requantize.h"
```
[the remark this answers is not in the document]
Comment (BalyshevArtem) [reply compiler/luci-interpreter/pal/cmsisnn/PALQuantize.h]: Done

**Review by m-bronnikov — APPROVED** [review]
LGTM, Thank you!

**Comment by jinevening** [comment]
LGTM

**Review by binarman — APPROVED** [review]
LGTM

**Pull request merged**